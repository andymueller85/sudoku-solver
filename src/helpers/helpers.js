export const GRID_SIZE = 9
export const possibleNums = '123456789'.split('')
export const allIndexes = Array.from({ length: GRID_SIZE }, (_, i) => i)
export const boxIndexes = [0, 3, 6]
export const miniGridIndexes = [0, 1, 2]

//************** General helper fns ****************/
export function isFilled(cell) {
  return cell !== '.'
}

export function getFilledCellCount(grid) {
  return grid.map(r => r.filter(c => isFilled(c))).flat().length
}

export function arrayIntersection(...arrays) {
  return arrays.reduce((acc, cur) => acc.filter(a => cur.includes(a)))
}

export function getColumnArray(grid, colNum) {
  return grid.map(r => r[colNum])
}

export function swapXY(grid) {
  return allIndexes.map(c => grid.map(r => r[c]))
}

export function allArraysAreEqual(...arrayOfArrays) {
  const firstArr = arrayOfArrays[0]

  return arrayOfArrays
    .slice(1)
    .every(
      a => a.length === firstArr.length && a.every((element, idx) => element === firstArr[idx])
    )
}

export function getUniqueArrays(...arrayOfArrays) {
  let uniqueArrays = [arrayOfArrays[0]]

  arrayOfArrays.slice(1).forEach(a => {
    if (uniqueArrays.every(u => !allArraysAreEqual(u, a))) {
      uniqueArrays.push(a)
    }
  })

  return uniqueArrays
}

export function gridHasAnyDeadEnds(grid) {
  return grid.some((r, rIdx) =>
    r.some(
      (_, cIdx) =>
        !isFilled(grid[rIdx][cIdx]) && getPossibleCellValues(grid, rIdx, cIdx).length === 0
    )
  )
}

export function getNextEmptyCellCoordinates(grid, rowNum, colNum) {
  let newRowNum = rowNum
  let newColNum = colNum
  const lastIndex = GRID_SIZE - 1

  if (rowNum === lastIndex && colNum === lastIndex) return undefined

  if (colNum === lastIndex) {
    newRowNum += 1
    newColNum = 0
  } else {
    newColNum += 1
  }

  if (!isFilled(grid[newRowNum][newColNum])) return [newRowNum, newColNum]

  return getNextEmptyCellCoordinates(grid, newRowNum, newColNum)
}

export function stringifyGrid(grid, isPossibleVals = false) {
  const vertSeparator = '|'
  const innerSeparator = `  ${vertSeparator}  `
  const rowBegin = `${vertSeparator}  `
  const rowEnd = `  ${vertSeparator}`
  const pad = rowBegin.length + rowEnd.length
  const horSeparator = '-'
  const boxSeparatorVert = String.fromCharCode(parseInt('275A', 16))
  const boxSeparatorHor = isPossibleVals ? horSeparator : '='

  function replaceCharAt(str, i, char) {
    return str.substring(0, i) + char + str.substring(i + 1)
  }

  let strGrid = horSeparator.repeat((GRID_SIZE - 1) * innerSeparator.length + GRID_SIZE + pad)

  for (let row = 0; row < GRID_SIZE; row++) {
    let logStr = grid[row].map(c => (isFilled(c) ? c : ' ')).join(innerSeparator)

    if (!isPossibleVals) {
      logStr = replaceCharAt(logStr, 15, boxSeparatorVert)
      logStr = replaceCharAt(logStr, 33, boxSeparatorVert)
    }

    strGrid += `\n${rowBegin}${logStr}${rowEnd}\n`
    strGrid += ([2, 5].includes(row) ? boxSeparatorHor : horSeparator).repeat(logStr.length + pad)
  }

  return strGrid
}

/* istanbul ignore next */
export function printGrid(grid) {
  console.log(stringifyGrid(grid))
}

export function seedGrid(input) {
  return input
    .split(/\r?\n/)
    .filter(r => r)
    .map(r => [...r].map(c => c))
}

export function getGridPossibleValues(grid) {
  return grid.map((r, rIdx) =>
    r.map((c, cIdx) => {
      if (isFilled(c)) return [c]

      const rowMissingNums = getRowMissingNums(grid, rIdx)
      const colMissingNums = getColumnMissingNums(grid, cIdx)
      const boxMissingNums = getBoxMissingNums(grid, getBoxTopLeft(rIdx), getBoxTopLeft(cIdx))

      return arrayIntersection(rowMissingNums, colMissingNums, boxMissingNums)
    })
  )
}

//************** Box helper fns ****************/
export function getBoxTopLeft(rowOrCol) {
  return rowOrCol - (rowOrCol % 3)
}

export function getBoxCurRow(topLeftRow, cell) {
  return topLeftRow + Math.floor(cell / 3)
}

export function getBoxCurColumn(topLeftCol, cell) {
  return topLeftCol + (cell % 3)
}

export function getTopLeftColumnForBoxNum(boxNum) {
  return 3 * (boxNum % 3)
}

export function getBoxTopLeftCoordinates(boxNum) {
  return [getBoxTopLeft(boxNum), getTopLeftColumnForBoxNum(boxNum)]
}

export function getBoxIndexes(topLeftRowOrColumn) {
  return [topLeftRowOrColumn, topLeftRowOrColumn + 1, topLeftRowOrColumn + 2]
}

export function flattenBox(grid, topLeftRowNum, topLeftColNum) {
  return grid
    .slice(topLeftRowNum, topLeftRowNum + 3)
    .map(r => r.slice(topLeftColNum, topLeftColNum + 3))
    .flat()
}

export function unflattenBox(boxArray) {
  return boxIndexes.map(i => boxArray.slice(i, i + 3))
}

export function flattenBoxes(grid) {
  return boxIndexes.map(r => boxIndexes.map(c => flattenBox(grid, r, c))).flat()
}

export function unflattenBoxes(grid) {
  let unflattenedGrid = Array.from({ length: GRID_SIZE }, () => Array.from({ length: GRID_SIZE }))

  grid.forEach((b, bIdx) => {
    const [topLeftRow, topLeftCol] = getBoxTopLeftCoordinates(bIdx)

    unflattenBox(b).forEach((r, rIdx) => {
      r.forEach((c, cIdx) => {
        unflattenedGrid[topLeftRow + rIdx][topLeftCol + cIdx] = c
      })
    })
  })

  return unflattenedGrid
}

export function rowNeighborsAreFilled(grid, curRow, curCol) {
  const topLeftCol = getBoxTopLeft(curCol)

  return grid[curRow]
    .map((cell, index) => ({ cell, index }))
    .slice(topLeftCol, topLeftCol + 3)
    .filter(({ index }) => index !== curCol)
    .every(({ cell }) => isFilled(cell))
}

export function columnNeighborsAreFilled(grid, curRow, curCol) {
  return rowNeighborsAreFilled(swapXY(grid), curCol, curRow)
}

export function rowNeighborsContainNumber(grid, curRow, num) {
  return getBoxIndexes(getBoxTopLeft(curRow))
    .filter(r => r !== curRow)
    .every(r => getFilledNums(grid[r]).includes(num))
}

export function columnNeighborsContainNumber(grid, curCol, num) {
  return getBoxIndexes(getBoxTopLeft(curCol))
    .filter(c => c !== curCol)
    .every(c => getColumnFilledNums(grid, c).includes(num))
}

export function getPossibleCellValues(grid, rowNum, colNum) {
  const rowMissingNums = getRowMissingNums(grid, rowNum)
  const colMissingNums = getColumnMissingNums(grid, colNum)
  const boxMissingNums = getBoxMissingNums(grid, getBoxTopLeft(rowNum), getBoxTopLeft(colNum))

  return arrayIntersection(rowMissingNums, colMissingNums, boxMissingNums)
}

//************** Filled Nums fns ****************/
export function getFilledNums(arr) {
  return arr.filter(c => isFilled(c))
}

export function getRowFilledNums(grid, rowNum) {
  return getFilledNums(grid[rowNum])
}

export function getColumnFilledNums(grid, colNum) {
  return getFilledNums(getColumnArray(grid, colNum))
}

export function getBoxFilledNums(grid, topLeftRowNum, topLeftColNum) {
  return getFilledNums(flattenBox(grid, topLeftRowNum, topLeftColNum))
}

//************** Missing Nums fns ****************/
export function getMissingNums(arr) {
  return possibleNums.filter(n => !getFilledNums(arr).includes(n))
}

export function getRowMissingNums(grid, rowNum) {
  return getMissingNums(grid[rowNum])
}

export function getColumnMissingNums(grid, colNum) {
  return getMissingNums(getColumnArray(grid, colNum))
}

export function getBoxMissingNums(grid, topLeftRowNum, topLeftColNum) {
  return getMissingNums(flattenBox(grid, topLeftRowNum, topLeftColNum))
}

//************** Missing Cells fns ****************/
export function getMissingCells(arr) {
  return arr
    .map((cell, index) => ({ cell, index }))
    .filter(({ cell }) => !isFilled(cell))
    .map(({ index }) => index)
}

export function getRowMissingCells(grid, rowNum) {
  return getMissingCells(grid[rowNum])
}

export function getColumnMissingCells(grid, colNum) {
  return getMissingCells(getColumnArray(grid, colNum))
}

export function getBoxMissingCells(grid, topLeftRowNum, topLeftColNum) {
  return getMissingCells(flattenBox(grid, topLeftRowNum, topLeftColNum))
}

//************** validity fns ****************/
export function isValid(arr) {
  const filledNums = getFilledNums(arr)

  return (
    allArraysAreEqual(filledNums, [...new Set(filledNums)]) &&
    filledNums.every(n => possibleNums.includes(n))
  )
}

export function rowIsValid(grid, rowNum) {
  return isValid(grid[rowNum])
}

export function everyRowIsValid(grid) {
  return grid.every((_, rowNum) => rowIsValid(grid, rowNum))
}

export function columnIsValid(grid, colNum) {
  return isValid(getColumnArray(grid, colNum))
}

export function everyColumnIsValid(grid) {
  return everyRowIsValid(swapXY(grid))
}

export function boxIsValid(grid, topLeftRowNum, topLeftColNum) {
  return isValid(flattenBox(grid, topLeftRowNum, topLeftColNum))
}

export function everyBoxIsValid(grid) {
  return everyRowIsValid(flattenBoxes(grid))
}

export function gridIsValid(grid) {
  return everyRowIsValid(grid) && everyColumnIsValid(grid) && everyBoxIsValid(grid)
}

//************** completeness fns ****************/
export function isComplete(arr) {
  return isValid(arr) && arr.length === GRID_SIZE && arr.every(c => isFilled(c))
}

export function rowIsComplete(grid, rowNum) {
  return isComplete(grid[rowNum])
}

export function columnIsComplete(grid, colNum) {
  return isComplete(getColumnArray(grid, colNum))
}

export function boxIsComplete(grid, topLeftRowNum, topLeftColNum) {
  return isComplete(flattenBox(grid, topLeftRowNum, topLeftColNum))
}

export function gridIsComplete(grid) {
  return gridIsValid(grid) && grid.every(r => r.every(c => isFilled(c)))
}
