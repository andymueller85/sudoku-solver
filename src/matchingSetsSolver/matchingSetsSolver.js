import lodash from 'lodash'
import {
  allIndexes,
  flattenBoxes,
  getUniqueArrays,
  GRID_SIZE,
  swapXY,
  unflattenBoxes
} from '../helpers/helpers.js'

const { cloneDeep } = lodash

export function findMatches(arr) {
  return arr.map((c, arrIdx, cells) => ({
    possibles: c,
    matches: cells
      .map((vals, idx) => ({ vals, idx }))
      .filter(
        cell =>
          arrIdx !== cell.idx &&
          cell.vals.length === c.length &&
          cell.vals.every((v, i) => v === c[i])
      )
      .map(({ idx }) => idx)
  }))
}

function whittlePossibles(matches, possiblesGrid) {
  let possibleValsGrid = cloneDeep(possiblesGrid)

  for (let n = 2; n < GRID_SIZE; n++) {
    matches.forEach((m, rowIdx) => {
      const matchesOfSizeN = m.filter(r => r.matches.length > 0 && r.possibles.length === n)

      if (matchesOfSizeN.length === n) {
        const matchingSets = getUniqueArrays(...matchesOfSizeN.map(m => m.possibles))

        matchingSets.forEach(possibles => {
          const setIndexes = [...new Set(matchesOfSizeN.map(m => m.matches).flat())]

          allIndexes
            .filter(n => !setIndexes.includes(n))
            .forEach(colIdx => {
              possibleValsGrid[rowIdx][colIdx] = possibleValsGrid[rowIdx][colIdx].filter(
                (v, _, a) => n > a.length || !possibles.includes(v)
              )
            })
        })
      }
    })
  }

  return possibleValsGrid
}

export function processRowMatchingSets(possibleValsGrid) {
  return whittlePossibles(possibleValsGrid.map(findMatches), possibleValsGrid)
}

export function processColumnMatchingSets(possibleValsGrid) {
  const swappedGrid = swapXY(possibleValsGrid)
  const columnMatches = swappedGrid.map(findMatches)

  return swapXY(whittlePossibles(columnMatches, swappedGrid))
}

export function processBoxMatchingSets(possibleValsGrid) {
  const flattenedBoxGrid = flattenBoxes(possibleValsGrid)
  const boxMatches = flattenedBoxGrid.map(findMatches)

  return unflattenBoxes(whittlePossibles(boxMatches, flattenedBoxGrid))
}

export function processMatchingSets(possibleValsGrid) {
  let myGrid = processBoxMatchingSets(possibleValsGrid)
  myGrid = processColumnMatchingSets(myGrid)
  myGrid = processRowMatchingSets(myGrid)

  return myGrid
}
