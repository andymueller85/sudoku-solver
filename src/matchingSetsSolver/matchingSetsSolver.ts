import lodash from 'lodash'
import {
  allIndexes,
  flattenBoxes,
  getUniqueArrays,
  GRID_SIZE,
  swapXY,
  unflattenBoxes
} from '../helpers/helpers'
import { Matches, CandidatesGrid, CandidatesRow } from '../types'

const { cloneDeep } = lodash

export function findMatches(arr: CandidatesRow): Array<Matches> {
  return arr.map((c, arrIdx, cells) => ({
    candidates: c,
    matches: cells
      .map((vals, idx) => ({ vals, idx }))
      .filter(
        ({ vals, idx }) =>
          arrIdx !== idx && vals.length === c.length && vals.every((v, i) => v === c[i])
      )
      .map(({ idx }) => idx)
  }))
}

export function whittleCandidates(
  matches: Array<Matches[]>,
  candidatesGrid: CandidatesGrid
): CandidatesGrid {
  let candidatesGridClone = cloneDeep(candidatesGrid)

  for (let n = 2; n < GRID_SIZE; n++) {
    matches.forEach((m, rowIdx: number) => {
      const matchesOfSizeN = m.filter(r => r.matches.length > 0 && r.candidates.length === n)

      if (matchesOfSizeN.length === n) {
        const matchingSets = getUniqueArrays(...matchesOfSizeN.map(m => m.candidates))

        matchingSets.forEach(candidates => {
          const setIndexes = [...new Set(matchesOfSizeN.map(m => m.matches).flat())]

          allIndexes
            .filter(n => !setIndexes.includes(n))
            .forEach(colIdx => {
              candidatesGridClone[rowIdx][colIdx] = candidatesGridClone[rowIdx][colIdx].filter(
                (v, _, a) => n > a.length || !candidates.includes(v)
              )
            })
        })
      }
    })
  }

  return candidatesGridClone
}

export function processRowMatchingSets(candidatesGrid: CandidatesGrid) {
  return whittleCandidates(candidatesGrid.map(findMatches), candidatesGrid)
}

export function processColumnMatchingSets(candidatesGrid: CandidatesGrid) {
  const swappedGrid = swapXY(candidatesGrid)
  const columnMatches = swappedGrid.map(findMatches)

  return swapXY(whittleCandidates(columnMatches, swappedGrid))
}

export function processBoxMatchingSets(candidatesGrid: CandidatesGrid) {
  const flattenedBoxGrid = flattenBoxes(candidatesGrid)
  const boxMatches = flattenedBoxGrid.map(findMatches)

  return unflattenBoxes(whittleCandidates(boxMatches, flattenedBoxGrid))
}

export function processMatchingSets(candidatesGrid: CandidatesGrid) {
  let myGrid = processBoxMatchingSets(candidatesGrid)
  myGrid = processColumnMatchingSets(myGrid)
  myGrid = processRowMatchingSets(myGrid)

  return myGrid
}
