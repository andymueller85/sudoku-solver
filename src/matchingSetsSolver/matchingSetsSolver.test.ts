import fs from 'fs'
import { getGridCandidates, seedGrid, stringifyGrid } from '../helpers/helpers'
import * as matchingSetsSolver from './matchingSetsSolver'

const fileInput = fs.readFileSync('./input.txt', 'utf8')
const grid = seedGrid(fileInput)
const candidatesGrid = getGridCandidates(grid)

describe('Matching sets solver functions', () => {
  describe('findMatches', () => {
    const { findMatches } = matchingSetsSolver

    test('should find all sets of matches for an array of candidates', () => {
      expect(findMatches(candidatesGrid[0])).toMatchSnapshot()
    })
  })

  describe('whittleCandidates', () => {
    const { whittleCandidates, findMatches } = matchingSetsSolver

    test('should process the matching sets, finding matches where the count of the matching sets equals the count of numbers in the set, eliminating those values from all other sets', () => {
      // just printing original for comparison
      expect(stringifyGrid(candidatesGrid, true)).toMatchInlineSnapshot(`
        "-------------------------------------------------------
        |  9  |  1  |  6  |  4  |  2,5  |  2,8  |  2,5  |  3  |  7  |
        -------------------------------------------------------------
        |  2  |  3,5  |  4  |  5,7,8  |  1,5,7  |  1,7,8,9  |  6  |  4,5,8  |  4,5,8,9  |
        ---------------------------------------------------------------------------------
        |  5  |  8  |  7  |  2,5  |  3  |  6  |  1  |  4,5  |  2,4,5,9  |
        -----------------------------------------------------------------
        |  3,6,7,8  |  9  |  5  |  2,3,6,7  |  1,2,4,6,7  |  1,2,4,7  |  2,7  |  7,8  |  1,2,8  |
        -----------------------------------------------------------------------------------------
        |  7  |  4  |  1  |  9  |  8  |  5  |  3  |  6  |  2  |
        -------------------------------------------------------
        |  3,6,7,8  |  2,3,6  |  2,6,8  |  2,3,6,7  |  1,2,6,7  |  1,2,7  |  4  |  9  |  1,2,5,8  |
        -------------------------------------------------------------------------------------------
        |  5,6  |  5,6  |  3  |  1  |  9  |  4,7  |  8  |  2  |  4,5  |
        ---------------------------------------------------------------
        |  1,5,6,8  |  2,5,6  |  9  |  2,5,6,7,8  |  2,4,5,6,7  |  2,4,7,8  |  5,7  |  4,5,7  |  3  |
        ---------------------------------------------------------------------------------------------
        |  4  |  7  |  2,8  |  2,5,8  |  2,5  |  3  |  5,9  |  1  |  6  |
        -----------------------------------------------------------------"
      `)

      const updatedCandidates = whittleCandidates(candidatesGrid.map(findMatches), candidatesGrid)

      expect(updatedCandidates[0][5]).toEqual(['8'])
      expect(updatedCandidates[6][8]).toEqual(['4'])
      expect(stringifyGrid(updatedCandidates, true)).toMatchInlineSnapshot(`
        "-------------------------------------------------------
        |  9  |  1  |  6  |  4  |  2,5  |  8  |  2,5  |  3  |  7  |
        -----------------------------------------------------------
        |  2  |  3,5  |  4  |  5,7,8  |  1,5,7  |  1,7,8,9  |  6  |  4,5,8  |  4,5,8,9  |
        ---------------------------------------------------------------------------------
        |  5  |  8  |  7  |  2,5  |  3  |  6  |  1  |  4,5  |  2,4,5,9  |
        -----------------------------------------------------------------
        |  3,6,7,8  |  9  |  5  |  2,3,6,7  |  1,2,4,6,7  |  1,2,4,7  |  2,7  |  7,8  |  1,2,8  |
        -----------------------------------------------------------------------------------------
        |  7  |  4  |  1  |  9  |  8  |  5  |  3  |  6  |  2  |
        -------------------------------------------------------
        |  3,6,7,8  |  2,3,6  |  2,6,8  |  2,3,6,7  |  1,2,6,7  |  1,2,7  |  4  |  9  |  1,2,5,8  |
        -------------------------------------------------------------------------------------------
        |  5,6  |  5,6  |  3  |  1  |  9  |  4,7  |  8  |  2  |  4  |
        -------------------------------------------------------------
        |  1,5,6,8  |  2,5,6  |  9  |  2,5,6,7,8  |  2,4,5,6,7  |  2,4,7,8  |  5,7  |  4,5,7  |  3  |
        ---------------------------------------------------------------------------------------------
        |  4  |  7  |  2,8  |  2,5,8  |  2,5  |  3  |  5,9  |  1  |  6  |
        -----------------------------------------------------------------"
      `)
    })
  })

  describe('processRowMatchingSets', () => {
    const { processRowMatchingSets } = matchingSetsSolver

    test('should process all row matching sets', () => {
      const updatedCandidates = processRowMatchingSets(candidatesGrid)

      expect(updatedCandidates[6][8]).toEqual(['4'])
      expect(stringifyGrid(updatedCandidates, true)).toMatchInlineSnapshot(`
        "-------------------------------------------------------
        |  9  |  1  |  6  |  4  |  2,5  |  8  |  2,5  |  3  |  7  |
        -----------------------------------------------------------
        |  2  |  3,5  |  4  |  5,7,8  |  1,5,7  |  1,7,8,9  |  6  |  4,5,8  |  4,5,8,9  |
        ---------------------------------------------------------------------------------
        |  5  |  8  |  7  |  2,5  |  3  |  6  |  1  |  4,5  |  2,4,5,9  |
        -----------------------------------------------------------------
        |  3,6,7,8  |  9  |  5  |  2,3,6,7  |  1,2,4,6,7  |  1,2,4,7  |  2,7  |  7,8  |  1,2,8  |
        -----------------------------------------------------------------------------------------
        |  7  |  4  |  1  |  9  |  8  |  5  |  3  |  6  |  2  |
        -------------------------------------------------------
        |  3,6,7,8  |  2,3,6  |  2,6,8  |  2,3,6,7  |  1,2,6,7  |  1,2,7  |  4  |  9  |  1,2,5,8  |
        -------------------------------------------------------------------------------------------
        |  5,6  |  5,6  |  3  |  1  |  9  |  4,7  |  8  |  2  |  4  |
        -------------------------------------------------------------
        |  1,5,6,8  |  2,5,6  |  9  |  2,5,6,7,8  |  2,4,5,6,7  |  2,4,7,8  |  5,7  |  4,5,7  |  3  |
        ---------------------------------------------------------------------------------------------
        |  4  |  7  |  2,8  |  2,5,8  |  2,5  |  3  |  5,9  |  1  |  6  |
        -----------------------------------------------------------------"
      `)
    })
  })

  describe('processColumnMatchingSets', () => {
    const { processColumnMatchingSets } = matchingSetsSolver

    test('should process all column matching sets', () => {
      const updatedCandidates = processColumnMatchingSets(candidatesGrid)

      expect(updatedCandidates[1][4]).toEqual('17'.split(''))
      expect(updatedCandidates[3][4]).toEqual('1467'.split(''))
      expect(updatedCandidates[5][4]).toEqual('167'.split(''))
      expect(updatedCandidates[7][4]).toEqual('467'.split(''))
      expect(stringifyGrid(updatedCandidates, true)).toMatchInlineSnapshot(`
        "-------------------------------------------------------
        |  9  |  1  |  6  |  4  |  2,5  |  2,8  |  2,5  |  3  |  7  |
        -------------------------------------------------------------
        |  2  |  3,5  |  4  |  5,7,8  |  1,7  |  1,7,8,9  |  6  |  4,5,8  |  4,5,8,9  |
        -------------------------------------------------------------------------------
        |  5  |  8  |  7  |  2,5  |  3  |  6  |  1  |  4,5  |  2,4,5,9  |
        -----------------------------------------------------------------
        |  3,6,7,8  |  9  |  5  |  2,3,6,7  |  1,4,6,7  |  1,2,4,7  |  2,7  |  7,8  |  1,2,8  |
        ---------------------------------------------------------------------------------------
        |  7  |  4  |  1  |  9  |  8  |  5  |  3  |  6  |  2  |
        -------------------------------------------------------
        |  3,6,7,8  |  2,3,6  |  2,6,8  |  2,3,6,7  |  1,6,7  |  1,2,7  |  4  |  9  |  1,2,5,8  |
        -----------------------------------------------------------------------------------------
        |  5,6  |  5,6  |  3  |  1  |  9  |  4,7  |  8  |  2  |  4,5  |
        ---------------------------------------------------------------
        |  1,5,6,8  |  2,5,6  |  9  |  2,5,6,7,8  |  4,6,7  |  2,4,7,8  |  5,7  |  4,5,7  |  3  |
        -----------------------------------------------------------------------------------------
        |  4  |  7  |  2,8  |  2,5,8  |  2,5  |  3  |  5,9  |  1  |  6  |
        -----------------------------------------------------------------"
      `)
    })
  })

  describe('processBoxMatchingSets', () => {
    const { processBoxMatchingSets } = matchingSetsSolver

    test('should process all box matching sets', () => {
      const updatedCandidates = processBoxMatchingSets(candidatesGrid)

      expect(updatedCandidates[0][5]).toEqual(['8'])
      expect(updatedCandidates[1][3]).toEqual('78'.split(''))
      expect(updatedCandidates[1][4]).toEqual('17'.split(''))
      expect(updatedCandidates[7][0]).toEqual('18'.split(''))
      expect(updatedCandidates[7][1]).toEqual(['2'])
      expect(stringifyGrid(updatedCandidates, true)).toMatchInlineSnapshot(`
        "-------------------------------------------------------
        |  9  |  1  |  6  |  4  |  2,5  |  8  |  2,5  |  3  |  7  |
        -----------------------------------------------------------
        |  2  |  3,5  |  4  |  7,8  |  1,7  |  1,7,8,9  |  6  |  4,5,8  |  4,5,8,9  |
        -----------------------------------------------------------------------------
        |  5  |  8  |  7  |  2,5  |  3  |  6  |  1  |  4,5  |  2,4,5,9  |
        -----------------------------------------------------------------
        |  3,6,7,8  |  9  |  5  |  2,3,6,7  |  1,2,4,6,7  |  1,2,4,7  |  2,7  |  7,8  |  1,2,8  |
        -----------------------------------------------------------------------------------------
        |  7  |  4  |  1  |  9  |  8  |  5  |  3  |  6  |  2  |
        -------------------------------------------------------
        |  3,6,7,8  |  2,3,6  |  2,6,8  |  2,3,6,7  |  1,2,6,7  |  1,2,7  |  4  |  9  |  1,2,5,8  |
        -------------------------------------------------------------------------------------------
        |  5,6  |  5,6  |  3  |  1  |  9  |  4,7  |  8  |  2  |  4,5  |
        ---------------------------------------------------------------
        |  1,8  |  2  |  9  |  2,5,6,7,8  |  2,4,5,6,7  |  2,4,7,8  |  5,7  |  4,5,7  |  3  |
        -------------------------------------------------------------------------------------
        |  4  |  7  |  2,8  |  2,5,8  |  2,5  |  3  |  5,9  |  1  |  6  |
        -----------------------------------------------------------------"
      `)
    })
  })

  describe('processMatchingSets', () => {
    const { processMatchingSets } = matchingSetsSolver
    test('should process all row, column and box matching sets', () => {
      const updatedCandidates = processMatchingSets(candidatesGrid)

      expect(stringifyGrid(updatedCandidates, true)).toMatchInlineSnapshot(`
        "-------------------------------------------------------
        |  9  |  1  |  6  |  4  |  2,5  |  8  |  2,5  |  3  |  7  |
        -----------------------------------------------------------
        |  2  |  3,5  |  4  |  7,8  |  1,7  |  1,7,8,9  |  6  |  4,5,8  |  4,5,8,9  |
        -----------------------------------------------------------------------------
        |  5  |  8  |  7  |  2,5  |  3  |  6  |  1  |  4,5  |  2,4,5,9  |
        -----------------------------------------------------------------
        |  3,6,7,8  |  9  |  5  |  2,3,6,7  |  1,4,6,7  |  1,2,4,7  |  2,7  |  7,8  |  1,2,8  |
        ---------------------------------------------------------------------------------------
        |  7  |  4  |  1  |  9  |  8  |  5  |  3  |  6  |  2  |
        -------------------------------------------------------
        |  3,6,7,8  |  2,3,6  |  2,6,8  |  2,3,6,7  |  1,6,7  |  1,2,7  |  4  |  9  |  1,2,5,8  |
        -----------------------------------------------------------------------------------------
        |  5,6  |  5,6  |  3  |  1  |  9  |  4,7  |  8  |  2  |  4  |
        -------------------------------------------------------------
        |  1,8  |  2  |  9  |  2,5,6,7,8  |  4,6,7  |  2,4,7,8  |  5,7  |  4,5,7  |  3  |
        ---------------------------------------------------------------------------------
        |  4  |  7  |  2,8  |  2,5,8  |  2,5  |  3  |  5,9  |  1  |  6  |
        -----------------------------------------------------------------"
      `)
    })
  })
})
