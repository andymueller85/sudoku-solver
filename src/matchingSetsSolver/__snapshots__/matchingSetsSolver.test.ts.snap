// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matching sets solver functions findMatches should find all sets of matches for an array of candidates 1`] = `
Array [
  Object {
    "candidates": Array [
      "9",
    ],
    "matches": Array [],
  },
  Object {
    "candidates": Array [
      "1",
    ],
    "matches": Array [],
  },
  Object {
    "candidates": Array [
      "6",
    ],
    "matches": Array [],
  },
  Object {
    "candidates": Array [
      "4",
    ],
    "matches": Array [],
  },
  Object {
    "candidates": Array [
      "2",
      "5",
    ],
    "matches": Array [
      6,
    ],
  },
  Object {
    "candidates": Array [
      "2",
      "8",
    ],
    "matches": Array [],
  },
  Object {
    "candidates": Array [
      "2",
      "5",
    ],
    "matches": Array [
      4,
    ],
  },
  Object {
    "candidates": Array [
      "3",
    ],
    "matches": Array [],
  },
  Object {
    "candidates": Array [
      "7",
    ],
    "matches": Array [],
  },
]
`;
